frontendGraph/src/main/java/com/virtuslab/gitmachete/frontend/graph/repository/RepositoryGraph.java

```diff
@@ -6,95 +6,113 @@
-        /* showLongEdges */ false);
+    GraphElementManager graphElementManager = new GraphElementManager(/* repositoryGraph */ this);
+    printElementGenerator = new PrintElementGeneratorImpl(/* graph */ this, graphElementManager);
   }
 
-  public Collection<? extends PrintElement> getPrintElements(int rowIndex) {
+  public Collection<? extends PrintElement> getPrintElements(@NonNegative int rowIndex) {
     return printElementGenerator.getPrintElements(rowIndex);
   }
 
   @SuppressWarnings("index:argument.type.incompatible")
-  public IGraphElement getGraphElement(int rowIndex) {
-    return elements.get(rowIndex);
+  public IGraphNode getGraphNode(@NonNegative int rowIndex) {
+    return nodes.get(rowIndex);
   }
 
-  @Override
   @NonNegative
   public int nodesCount() {
-    return elements.size();
-  }
-
-  @Override
-  public GraphNode getGraphNode(int nodeIndex) {
-    return new GraphNode(nodeIndex);
+    return nodes.size();
   }
 
-  @Override
   public int getNodeId(int nodeIndex) {
     assert nodeIndex >= 0 && nodeIndex < nodesCount() : "Bad nodeIndex: " + nodeIndex;
     return nodeIndex;
   }
 
-  @Override
-  @Nullable
-  public Integer getNodeIndex(int nodeId) {
-    if (nodeId >= 0 && nodeId < nodesCount()) {
-      return nodeId;
-    }
-    return null;
-  }
-
-  @Override
-  public java.util.List<GraphEdge> getAdjacentEdges(int nodeIndex, EdgeFilter filter) {
+  /**
+   * Adjacent edges are the edges that are visible in a row and directly connected to the node (branch/commit node)
+   * of this row. See {@link RepositoryGraph#getVisibleEdgesWithPositions} for more details.
+   *
+   *  @return list of adjacent edges in a given node index
+   */
+  public java.util.List<GraphEdge> getAdjacentEdges(@NonNegative int nodeIndex, EdgeFilter filter) {
     if (filter == EdgeFilter.SPECIAL) {
       return Collections.emptyList();
     }
 
     java.util.List<GraphEdge> adjacentEdges = new SmartList<>();
     @SuppressWarnings("index:argument.type.incompatible")
-    IGraphElement currentElement = elements.get(nodeIndex);
-
-    if (filter.downNormal && nodeIndex < elements.size() - 1) {
-      currentElement.getDownElementIndexes().stream()
-          .map(i -> GraphEdge.createNormalEdge(nodeIndex, i, GraphEdgeType.USUAL)).forEach(adjacentEdges::add);
-    }
+    IGraphNode currentNode = nodes.get(nodeIndex);
 
     if (filter.upNormal && nodeIndex > 0) {
-      int upIndex = currentElement.getUpElementIndex();
+      int upIndex = currentNode.getPrevSiblingNodeIndex();
       if (upIndex >= 0) {
         adjacentEdges.add(GraphEdge.createNormalEdge(nodeIndex, upIndex, GraphEdgeType.USUAL));
       }
     }
 
+    if (filter.downNormal && nodeIndex < nodes.size() - 1) {
+      Integer nextSiblingNodeIndex = currentNode.getNextSiblingNodeIndex();
+      if (nextSiblingNodeIndex != null) {
+        adjacentEdges.add(GraphEdge.createNormalEdge(nodeIndex, nextSiblingNodeIndex, GraphEdgeType.USUAL));
```

Comment:
Stub-annotate `int` params `GraphEdge#createNormalEdge` for lower bounds (`@NonNegative`?)